[Java] Fury codegen error for package access level class

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/alipay/issues) and found no similar issues.


### Version

445bfa7e9214a4626f8f8f829a8464fb7ce4f622

### Component(s)

Java

### Minimal reproduce step

```java

 Fury fury = Fury.builder().withRefTracking(true).requireClassRegistration(false).build();
    HashBasedTable<Object, Object, Object> table = HashBasedTable.create(2, 4);
    table.put("r", "c", 100);
```

### What did you expect to see?

No error

### What did you see instead?

```java
java.lang.RuntimeException: Create sequential serializer failed, 
class: class com.google.common.collect.HashBasedTable

	at io.fury.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:48)
	at io.fury.resolver.ClassResolver.lambda$getObjectSerializerClass$2(ClassResolver.java:962)
	at io.fury.builder.JITContext.registerSerializerJITCallback(JITContext.java:132)
	at io.fury.resolver.ClassResolver.getObjectSerializerClass(ClassResolver.java:960)
	at io.fury.resolver.ClassResolver.getSerializerClass(ClassResolver.java:894)
	at io.fury.resolver.ClassResolver.getSerializerClass(ClassResolver.java:791)
	at io.fury.resolver.ClassResolver.createSerializer(ClassResolver.java:1182)
	at io.fury.resolver.ClassResolver.getOrUpdateClassInfo(ClassResolver.java:1114)
	at io.fury.Fury.writeRef(Fury.java:342)
	at io.fury.Fury.write(Fury.java:319)
	at io.fury.Fury.serialize(Fury.java:255)
	at io.fury.Fury.serialize(Fury.java:208)
	at io.fury.FuryTestBase.serDeCheckSerializer(FuryTestBase.java:177)
	at io.fury.FuryTest.testPkgAccessLevelParentClass(FuryTest.java:620)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:136)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:658)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:219)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:923)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:192)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.testng.TestRunner.privateRun(TestRunner.java:808)
	at org.testng.TestRunner.run(TestRunner.java:603)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:429)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:423)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:383)
	at org.testng.SuiteRunner.run(SuiteRunner.java:326)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1249)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1169)
	at org.testng.TestNG.runSuites(TestNG.java:1092)
	at org.testng.TestNG.run(TestNG.java:1060)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:105)
Caused by: io.fury.codegen.CodegenException: Compile error: 
com.google.common.collect.StandardTableFuryAccessor_622488023_635611994:
/* 0001 */ package com.google.common.collect;
/* 0002 */ 
/* 0003 */ public final class StandardTableFuryAccessor_622488023_635611994 {
/* 0004 */ 
/* 0005 */   public static java.util.Map backingMap(Object obj) {
/* 0006 */       return obj.backingMap;
/* 0007 */   }
/* 0008 */ 
/* 0009 */   public static void backingMap(Object obj, java.util.Map fieldValue) {
/* 0010 */       obj.backingMap = fieldValue;
/* 0011 */   }
/* 0012 */ 
/* 0013 */   public static com.google.common.base.Supplier factory(Object obj) {
/* 0014 */       return obj.factory;
/* 0015 */   }
/* 0016 */ 
/* 0017 */   public static void factory(Object obj, com.google.common.base.Supplier fieldValue) {
/* 0018 */       obj.factory = fieldValue;
/* 0019 */   }
/* 0020 */ 
/* 0021 */ }
	at io.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:126)
	at io.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:64)
	at io.fury.builder.AccessorHelper.defineAccessorClass(AccessorHelper.java:163)
	at io.fury.builder.AccessorHelper.defineAccessor(AccessorHelper.java:212)
	at io.fury.builder.CodecBuilder.getFieldValue(CodecBuilder.java:207)
	at io.fury.builder.ObjectCodecBuilder.lambda$serializeGroup$9f8a61a4$1(ObjectCodecBuilder.java:185)
	at io.fury.builder.ObjectCodecBuilder.serializeGroup(ObjectCodecBuilder.java:194)
	at io.fury.builder.ObjectCodecBuilder.buildEncodeExpression(ObjectCodecBuilder.java:158)
	at io.fury.builder.BaseObjectCodecBuilder.genCode(BaseObjectCodecBuilder.java:204)
	at io.fury.codegen.CompileUnit.getCode(CompileUnit.java:54)
	at io.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:72)
	at io.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:64)
	at io.fury.codegen.CodeGenerator.compile(CodeGenerator.java:144)
	at io.fury.builder.CodecUtils.loadOrGenCodecClass(CodecUtils.java:91)
	at io.fury.builder.CodecUtils.loadOrGenObjectCodecClass(CodecUtils.java:42)
	at io.fury.serializer.CodegenSerializer.loadCodegenSerializer(CodegenSerializer.java:45)
	... 40 more
Caused by: org.codehaus.commons.compiler.CompileException: File 'com/google/common/collect/StandardTableFuryAccessor_622488023_635611994.java', Line 6, Column 18: "backingMap" is neither a method, a field, nor a member class of "java.lang.Object"
	at org.codehaus.janino.UnitCompiler.compileError(UnitCompiler.java:13080)
	at org.codehaus.janino.UnitCompiler.reclassifyName(UnitCompiler.java:8896)
	at org.codehaus.janino.UnitCompiler.reclassify(UnitCompiler.java:8698)
	at org.codehaus.janino.UnitCompiler.getConstantValue2(UnitCompiler.java:5992)
	at org.codehaus.janino.UnitCompiler.access$11000(UnitCompiler.java:237)
	at org.codehaus.janino.UnitCompiler$19$1.visitAmbiguousName(UnitCompiler.java:5944)
	at org.codehaus.janino.Java$AmbiguousName.accept(Java.java:4603)
	at org.codehaus.janino.UnitCompiler$19.visitLvalue(UnitCompiler.java:5943)
	at org.codehaus.janino.Java$Lvalue.accept(Java.java:4528)
	at org.codehaus.janino.UnitCompiler.getConstantValue(UnitCompiler.java:5939)
	at org.codehaus.janino.UnitCompiler.compileGetValue(UnitCompiler.java:5876)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:2849)
	at org.codehaus.janino.UnitCompiler.access$2800(UnitCompiler.java:237)
	at org.codehaus.janino.UnitCompiler$6.visitReturnStatement(UnitCompiler.java:1591)
	at org.codehaus.janino.UnitCompiler$6.visitReturnStatement(UnitCompiler.java:1576)
	at org.codehaus.janino.Java$ReturnStatement.accept(Java.java:3888)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:1576)
	at org.codehaus.janino.UnitCompiler.compileStatements(UnitCompiler.java:1662)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:3665)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:3330)
	at org.codehaus.janino.UnitCompiler.compileDeclaredMethods(UnitCompiler.java:1448)
	at org.codehaus.janino.UnitCompiler.compileDeclaredMethods(UnitCompiler.java:1421)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:830)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:443)
	at org.codehaus.janino.UnitCompiler.access$400(UnitCompiler.java:237)
	at org.codehaus.janino.UnitCompiler$3.visitPackageMemberClassDeclaration(UnitCompiler.java:423)
	at org.codehaus.janino.UnitCompiler$3.visitPackageMemberClassDeclaration(UnitCompiler.java:419)
	at org.codehaus.janino.Java$PackageMemberClassDeclaration.accept(Java.java:1688)
	at org.codehaus.janino.UnitCompiler.compile(UnitCompiler.java:419)
	at org.codehaus.janino.UnitCompiler.compile2(UnitCompiler.java:393)
	at org.codehaus.janino.UnitCompiler.access$000(UnitCompiler.java:237)
	at org.codehaus.janino.UnitCompiler$2.visitCompilationUnit(UnitCompiler.java:364)
	at org.codehaus.janino.UnitCompiler$2.visitCompilationUnit(UnitCompiler.java:362)
	at org.codehaus.janino.Java$CompilationUnit.accept(Java.java:371)
	at org.codehaus.janino.UnitCompiler.compileUnit(UnitCompiler.java:362)
	at org.codehaus.janino.Compiler.compile2(Compiler.java:271)
	at org.codehaus.janino.Compiler.compile(Compiler.java:213)
	at io.fury.codegen.JaninoUtils.toBytecode(JaninoUtils.java:106)
	... 55 more

```

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/java/fury-core/src/main/java/io/fury/builder/AccessorHelper.java
+++ b/java/fury-core/src/main/java/io/fury/builder/AccessorHelper.java
@@ -16,6 +16,8 @@
 
 package io.fury.builder;
 
+import static io.fury.codegen.CodeGenerator.sourcePkgLevelAccessible;
+
 import io.fury.codegen.CodeGenerator;
 import io.fury.codegen.CodegenContext;
 import io.fury.codegen.CompileUnit;
@@ -78,50 +80,53 @@ public class AccessorHelper {
     // filter out super classes
     Collection<Descriptor> descriptors = Descriptor.getAllDescriptorsMap(beanClass, false).values();
     for (Descriptor descriptor : descriptors) {
-      if (!Modifier.isPrivate(descriptor.getModifiers())) {
-        {
-          String methodName = descriptor.getName();
-          String codeBody =
+      if (Modifier.isPrivate(descriptor.getModifiers())) {
+        continue;
+      }
+      boolean accessible = sourcePkgLevelAccessible(descriptor.getRawType());
+      {
+        // getter
+        String methodName = descriptor.getName();
+        String codeBody;
+        Class<?> returnType = accessible ? descriptor.getRawType() : Object.class;
+        if (isRecord) {
+          codeBody =
+              StringUtils.format(
+                  "return ${obj}.${fieldName}();",
+                  "obj",
+                  OBJ_NAME,
+                  "fieldName",
+                  descriptor.getName());
+        } else {
+          codeBody =
               StringUtils.format(
                   "return ${obj}.${fieldName};",
                   "obj",
                   OBJ_NAME,
                   "fieldName",
                   descriptor.getName());
-          Class<?> returnType = descriptor.getRawType();
-          ctx.addStaticMethod(methodName, codeBody, returnType, beanClass, OBJ_NAME);
         }
-        {
-          String methodName = descriptor.getName();
-          String codeBody =
-              StringUtils.format(
-                  "${obj}.${fieldName} = ${fieldValue};",
-                  "obj",
-                  OBJ_NAME,
-                  "fieldName",
-                  descriptor.getName(),
-                  "fieldValue",
-                  FIELD_VALUE);
-          ctx.addStaticMethod(
-              methodName,
-              codeBody,
-              void.class,
-              beanClass,
-              OBJ_NAME,
-              descriptor.getRawType(),
-              FIELD_VALUE);
-        }
-      } else if (isRecord) {
+        ctx.addStaticMethod(methodName, codeBody, returnType, beanClass, OBJ_NAME);
+      }
+      if (accessible) {
         String methodName = descriptor.getName();
         String codeBody =
             StringUtils.format(
-                "return ${obj}.${fieldName}();",
+                "${obj}.${fieldName} = ${fieldValue};",
                 "obj",
                 OBJ_NAME,
                 "fieldName",
-                descriptor.getName());
-        Class<?> returnType = descriptor.getRawType();
-        ctx.addStaticMethod(methodName, codeBody, returnType, beanClass, OBJ_NAME);
+                descriptor.getName(),
+                "fieldValue",
+                FIELD_VALUE);
+        ctx.addStaticMethod(
+            methodName,
+            codeBody,
+            void.class,
+            beanClass,
+            OBJ_NAME,
+            descriptor.getRawType(),
+            FIELD_VALUE);
       }
       // getter/setter may lose some inner state of an object, so we set them to null to avoid
       // creating getter/setter accessor.
@@ -194,26 +199,39 @@ public class AccessorHelper {
     }
   }
 
+  /** Should be invoked only when {@link #defineAccessor} returns true. */
   public static Class<?> getAccessorClass(Field field) {
     Class<?> beanClass = field.getDeclaringClass();
     return getAccessorClass(beanClass);
   }
 
+  /** Should be invoked only when {@link #defineAccessor} returns true. */
   public static Class<?> getAccessorClass(Method method) {
     Class<?> beanClass = method.getDeclaringClass();
     return getAccessorClass(beanClass);
   }
 
   public static boolean defineAccessor(Field field) {
-    if (ReflectionUtils.isPrivate(field.getType())) {
+    Class<?> beanClass = field.getDeclaringClass();
+    return defineAccessorClass(beanClass);
+  }
+
+  public static boolean defineAccessor(Method method) {
+    Class<?> beanClass = method.getDeclaringClass();
+    return defineAccessorClass(beanClass);
+  }
+
+  public static boolean defineSetter(Field field) {
+    if (ReflectionUtils.isPrivate(field.getType()) || !sourcePkgLevelAccessible(field.getType())) {
       return false;
     }
     Class<?> beanClass = field.getDeclaringClass();
     return defineAccessorClass(beanClass);
   }
 
-  public static boolean defineAccessor(Method method) {
-    if (ReflectionUtils.isPrivate(method.getReturnType())) {
+  public static boolean defineSetter(Method method) {
+    if (ReflectionUtils.isPrivate(method.getReturnType())
+        || !sourcePkgLevelAccessible(method.getReturnType())) {
       return false;
     }
     Class<?> beanClass = method.getDeclaringClass();

--- a/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/BaseObjectCodecBuilder.java
@@ -17,7 +17,6 @@
 package io.fury.builder;
 
 import static io.fury.codegen.CodeGenerator.getPackage;
-import static io.fury.codegen.CodeGenerator.sourceAccessible;
 import static io.fury.codegen.Expression.Invoke.inlineInvoke;
 import static io.fury.codegen.Expression.Reference.fieldRef;
 import static io.fury.codegen.ExpressionOptimizer.invokeGenerated;
@@ -530,7 +529,7 @@ public abstract class BaseObjectCodecBuilder extends CodecBuilder {
   }
 
   protected Expression getClassExpr(Class<?> cls) {
-    if (sourceAccessible(cls)) {
+    if (sourcePublicAccessible(cls)) {
       return Literal.ofClass(cls);
     } else {
       return new StaticInvoke(

--- a/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
@@ -121,16 +121,16 @@ public abstract class CodecBuilder {
   /** Returns an expression that serialize java bean of type {@link CodecBuilder#beanClass}. */
   public abstract Expression buildEncodeExpression();
 
+  protected boolean sourcePublicAccessible(Class<?> cls) {
+    return ctx.sourcePublicAccessible(cls);
+  }
+
   protected Expression tryInlineCast(Expression expression, TypeToken<?> targetType) {
     return tryCastIfPublic(expression, targetType, true);
   }
 
   protected Expression tryCastIfPublic(Expression expression, TypeToken<?> targetType) {
     return tryCastIfPublic(expression, targetType, false);
-  }
-
-  protected boolean sourcePublicAccessible(Class<?> cls) {
-    return ctx.sourcePublicAccessible(cls);
   }
 
   protected Expression tryCastIfPublic(

--- a/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
+++ b/java/fury-core/src/main/java/io/fury/builder/CodecBuilder.java
@@ -16,7 +16,6 @@
 
 package io.fury.builder;
 
-import static io.fury.codegen.CodeGenerator.sourceAccessible;
 import static io.fury.codegen.Expression.Invoke.inlineInvoke;
 import static io.fury.type.TypeUtils.CLASS_TYPE;
 import static io.fury.type.TypeUtils.OBJECT_ARRAY_TYPE;
@@ -130,6 +129,10 @@ public abstract class CodecBuilder {
     return tryCastIfPublic(expression, targetType, false);
   }
 
+  protected boolean sourcePublicAccessible(Class<?> cls) {
+    return ctx.sourcePublicAccessible(cls);
+  }
+
   protected Expression tryCastIfPublic(
       Expression expression, TypeToken<?> targetType, boolean inline) {
     Class<?> rawType = getRawType(targetType);
@@ -138,7 +141,7 @@ public abstract class CodecBuilder {
       return expression;
     }
     if (inline) {
-      if (sourceAccessible(rawType)) {
+      if (sourcePublicAccessible(rawType)) {
         return new Cast(expression, targetType);
       } else {
         return new Cast(expression, ReflectionUtils.getPublicSuperType(TypeToken.of(rawType)));
@@ -150,7 +153,8 @@ public abstract class CodecBuilder {
   protected Expression tryCastIfPublic(
       Expression expression, TypeToken<?> targetType, String valuePrefix) {
     Class<?> rawType = getRawType(targetType);
-    if (sourceAccessible(rawType) && !expression.type().wrap().isSubtypeOf(targetType.wrap())) {
+    if (sourcePublicAccessible(rawType)
+        && !expression.type().wrap().isSubtypeOf(targetType.wrap())) {
       return new Cast(expression, targetType, valuePrefix);
     }
     return expression;
@@ -185,15 +189,15 @@ public abstract class CodecBuilder {
   protected Expression getFieldValue(Expression inputBeanExpr, Descriptor descriptor) {
     TypeToken<?> fieldType = descriptor.getTypeToken();
     Class<?> rawType = descriptor.getRawType();
-    if (!sourceAccessible(rawType)) {
-      fieldType = OBJECT_TYPE;
-    }
     String fieldName = descriptor.getName();
     if (isRecord) {
       return getRecordFieldValue(inputBeanExpr, descriptor);
     }
     if (duplicatedFields.contains(fieldName) || !Modifier.isPublic(beanClass.getModifiers())) {
       return unsafeAccessField(inputBeanExpr, beanClass, descriptor);
+    }
+    if (!sourcePublicAccessible(rawType)) {
+      fieldType = OBJECT_TYPE;
     }
     // public field or non-private non-java field access field directly.
     if (Modifier.isPublic(descriptor.getModifiers())) {
@@ -230,6 +234,9 @@ public abstract class CodecBuilder {
 
   private Expression getRecordFieldValue(Expression inputBeanExpr, Descriptor descriptor) {
     TypeToken<?> fieldType = descriptor.getTypeToken();
+    if (!sourcePublicAccessible(descriptor.getRawType())) {
+      fieldType = OBJECT_TYPE;
+    }
     String fieldName = descriptor.getName();
     if (Modifier.isPublic(beanClass.getModifiers())) {
       Preconditions.checkNotNull(descriptor.getReadMethod());
@@ -256,9 +263,7 @@ public abstract class CodecBuilder {
       }
       if (!fieldType.isPrimitive()) {
         Expression v = inlineInvoke(ref, methodInfo.f1, OBJECT_TYPE, fieldNullable, inputBeanExpr);
-        TypeToken<?> publicSuperType =
-            ReflectionUtils.getPublicSuperType(descriptor.getTypeToken());
-        return new Cast(v, publicSuperType, fieldName);
+        return tryCastIfPublic(v, descriptor.getTypeToken(), fieldName);
       } else {
         return new Invoke(ref, methodInfo.f1, fieldType, fieldNullable, inputBeanExpr);
       }
@@ -296,8 +301,7 @@ public abstract class CodecBuilder {
               fieldNullable,
               inputObject,
               fieldOffsetExpr);
-      TypeToken<?> publicSuperType = ReflectionUtils.getPublicSuperType(descriptor.getTypeToken());
-      return tryCastIfPublic(getObj, publicSuperType, fieldName);
+      return tryCastIfPublic(getObj, descriptor.getTypeToken(), fieldName);
     }
   }
 
@@ -337,7 +341,7 @@ public abstract class CodecBuilder {
     if (value instanceof Inlineable) {
       ((Inlineable) value).inline();
     }
-    if (duplicatedFields.contains(fieldName) || !sourceAccessible(beanClass)) {
+    if (duplicatedFields.contains(fieldName) || !sourcePublicAccessible(beanClass)) {
       return unsafeSetField(bean, d, value);
     }
     if (!Modifier.isFinal(d.getModifiers()) && Modifier.isPublic(d.getModifiers())) {
@@ -346,7 +350,7 @@ public abstract class CodecBuilder {
       return new Invoke(bean, d.getWriteMethod().getName(), value);
     } else {
       if (!Modifier.isFinal(d.getModifiers()) && !Modifier.isPrivate(d.getModifiers())) {
-        if (AccessorHelper.defineAccessor(d.getField())) {
+        if (AccessorHelper.defineSetter(d.getField())) {
           Class<?> accessorClass = AccessorHelper.getAccessorClass(d.getField());
           if (!value.type().equals(d.getTypeToken())) {
             value = new Cast(value, d.getTypeToken());
@@ -356,7 +360,7 @@ public abstract class CodecBuilder {
         }
       }
       if (d.getWriteMethod() != null && !Modifier.isPrivate(d.getWriteMethod().getModifiers())) {
-        if (AccessorHelper.defineAccessor(d.getWriteMethod())) {
+        if (AccessorHelper.defineSetter(d.getWriteMethod())) {
           Class<?> accessorClass = AccessorHelper.getAccessorClass(d.getWriteMethod());
           if (!value.type().equals(d.getTypeToken())) {
             value = new Cast(value, d.getTypeToken());
@@ -452,7 +456,7 @@ public abstract class CodecBuilder {
   /** Returns an Expression that create a new java object of type {@link CodecBuilder#beanClass}. */
   protected Expression newBean() {
     // TODO allow default access-level class.
-    if (sourceAccessible(beanClass)) {
+    if (sourcePublicAccessible(beanClass)) {
       return new Expression.NewInstance(beanType);
     } else {
       return new StaticInvoke(Platform.class, "newInstance", OBJECT_TYPE, beanClassExpr());

--- a/java/fury-core/src/main/java/io/fury/codegen/CodeGenerator.java
+++ b/java/fury-core/src/main/java/io/fury/codegen/CodeGenerator.java
@@ -479,12 +479,23 @@ public class CodeGenerator {
     return sb;
   }
 
-  /** Returns true if class is accessible from source. */
-  public static boolean sourceAccessible(Class<?> clz) {
+  /** Returns true if class is public accessible from source. */
+  public static boolean sourcePublicAccessible(Class<?> clz) {
     if (clz.isPrimitive()) {
       return true;
     }
-    if (!ReflectionUtils.isPublic(clz) || clz.getCanonicalName() == null) {
+    if (!ReflectionUtils.isPublic(clz)) {
+      return false;
+    }
+    return sourcePkgLevelAccessible(clz);
+  }
+
+  /** Returns true if class is package level accessible from source. */
+  public static boolean sourcePkgLevelAccessible(Class<?> clz) {
+    if (clz.isPrimitive()) {
+      return true;
+    }
+    if (clz.getCanonicalName() == null) {
       return false;
     }
     // Scala may produce class name like: xxx.SomePackageObject.package$SomeClass

--- a/java/fury-core/src/main/java/io/fury/codegen/CodegenContext.java
+++ b/java/fury-core/src/main/java/io/fury/codegen/CodegenContext.java
@@ -275,7 +275,7 @@ public class CodegenContext {
    *     return canonical name otherwise.
    */
   public String type(Class<?> clz) {
-    if (!CodeGenerator.sourceAccessible(clz)) {
+    if (!sourcePkgLevelAccessible(clz)) {
       return "Object";
     }
     if (clz.isArray()) {
@@ -674,5 +674,19 @@ public class CodegenContext {
     } else {
       return code;
     }
+  }
+
+  private final Map<Class<?>, Boolean> sourcePublicAccessibleCache = new HashMap<>();
+  private final Map<Class<?>, Boolean> sourcePkgLevelAccessibleCache = new HashMap<>();
+
+  /** Returns true if class is public accessible from source. */
+  public boolean sourcePublicAccessible(Class<?> clz) {
+    return sourcePublicAccessibleCache.computeIfAbsent(clz, CodeGenerator::sourcePublicAccessible);
+  }
+
+  /** Returns true if class is package level accessible from source. */
+  public boolean sourcePkgLevelAccessible(Class<?> clz) {
+    return sourcePkgLevelAccessibleCache.computeIfAbsent(
+        clz, CodeGenerator::sourcePkgLevelAccessible);
   }
 }

--- a/java/fury-core/src/test/java/io/fury/FuryTest.java
+++ b/java/fury-core/src/test/java/io/fury/FuryTest.java
@@ -21,6 +21,7 @@ import static org.testng.Assert.assertNull;
 import static org.testng.Assert.assertThrows;
 import static org.testng.Assert.assertTrue;
 
+import com.google.common.collect.HashBasedTable;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import io.fury.annotation.Ignore;
@@ -609,5 +610,13 @@ public class FuryTest extends FuryTestBase {
     } catch (StackOverflowError e) {
       Assert.assertTrue(e.getMessage().contains("reference"));
     }
+  }
+
+  @Test
+  public void testPkgAccessLevelParentClass() {
+    Fury fury = Fury.builder().withRefTracking(true).requireClassRegistration(false).build();
+    HashBasedTable<Object, Object, Object> table = HashBasedTable.create(2, 4);
+    table.put("r", "c", 100);
+    serDeCheckSerializer(fury, table, "Codec");
   }
 }

--- a/java/fury-core/src/test/java/io/fury/builder/AccessorHelperTest.java
+++ b/java/fury-core/src/test/java/io/fury/builder/AccessorHelperTest.java
@@ -22,6 +22,7 @@ import static org.testng.Assert.assertSame;
 import static org.testng.Assert.assertTrue;
 
 import io.fury.TestUtils;
+import io.fury.builder.pkg.AccessLevelClass;
 import io.fury.test.bean.Foo;
 import io.fury.test.bean.Struct;
 import io.fury.type.Descriptor;
@@ -45,7 +46,8 @@ public class AccessorHelperTest {
 
   @Test
   public void genCode() {
-    AccessorHelper.genCode(A.class);
+    System.out.println(AccessorHelper.genCode(A.class));
+    ;
   }
 
   @Test
@@ -63,6 +65,20 @@ public class AccessorHelperTest {
     assertTrue(AccessorHelper.defineAccessor(A.class.getDeclaredField("f1")));
     assertTrue(AccessorHelper.defineAccessor(A.class.getDeclaredMethod("getF1")));
     assertSame(AccessorHelper.getAccessorClass(A.class), accessorClass);
+  }
+
+  @Data
+  static class PkgAccessAccessorClass {
+    protected String f1;
+    String f2;
+  }
+
+  @Test
+  public void definePkgAccessAccessorClass() {
+    assertTrue(AccessorHelper.defineAccessorClass(PkgAccessAccessorClass.class));
+    assertTrue(AccessorHelper.defineAccessorClass(AccessLevelClass.class));
+    Method[] methods = AccessorHelper.getAccessorClass(AccessLevelClass.class).getDeclaredMethods();
+    assertEquals(methods.length, 4);
   }
 
   @Test

--- a/java/fury-core/src/test/java/io/fury/builder/pkg/AccessLevelClass.java
+++ b/java/fury-core/src/test/java/io/fury/builder/pkg/AccessLevelClass.java
@@ -1,0 +1,31 @@
+/*
+ * Copyright 2023 The Fury Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.fury.builder.pkg;
+
+public class AccessLevelClass {
+  static class PkgLevelClass {
+    int f1;
+  }
+
+  static class PrivateLevelClass {
+    int f1;
+  }
+
+  PkgLevelClass f1;
+  PrivateLevelClass f2;
+  private PrivateLevelClass f3;
+}

--- a/scala/src/test/scala/io/fury/serializer/ScalaTest.scala
+++ b/scala/src/test/scala/io/fury/serializer/ScalaTest.scala
@@ -41,6 +41,7 @@ class ScalaTest extends AnyWordSpec with Matchers {
   "serialize/deserialize package object in app" in {
     // If we move code in main here, we can't reproduce https://github.com/alipay/fury/issues/1165.
     PkgObjectMain.main(Array())
+    PkgObjectMain2.main(Array())
   }
 }
 

